fms_extras/utils/generation.py

```diff
@@ -0,0 +1,141 @@
+import functools
+import time
+from typing import Any, Callable, List, MutableMapping, Optional, Union
+
+import torch
+import torch.nn.functional as F
+
+from fms_extras.utils.cache.paged import PagedKVCacheManager
+
+
+def paged_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids: torch.Tensor,
+    kv_cache_manager: PagedKVCacheManager,
+    max_seq_len: int = 2048,
+    max_new_tokens: int = 256,
+    temperature: float = 1.0,
+    top_k: int = 10,
+    do_sample: bool = True,
+    decode_model: Optional[Union[Callable, torch.nn.Module]] = None,
+    # todo: This is a WIP to enable cudagraphs, currently its only for batch_size=1
+    cudagraphs: bool = False,
+):
+    """
+    A trivial generate function that can be used for validation/testing generation using paged attention
+
+    Args:
+        model: Callable or nn.Module
+            A function or nn.Module that takes a batch of input_ids and returns logits
+        kv_cache_manager: PagedKVCacheManager
+            the paged KVCacheManager that handles management of the kv-cache for paged attention
+        max_seq_len: int
+            the max sequence length of the model
+        max_new_tokens: int
+            max tokens to generate
+        temperature: float
+            temperature of softmax when sampling
+        top_k: int
+            only search among top k tokens
+        do_sample: bool
+            multinomial sampling. False for greedy.
+        decode_model: Callable or nn.Module, optional
+            a model to used specifically for decode step. If not given, the model input will be used for decode step
+
+    Returns:
+    Tuple[torch.Tensor, int, int]
+        the resulting output tokens, the number of new tokens generated, and the time it took per token in seconds
```

Comment:
the third term returned is `(end_time - start_time)` so I think this should be float seconds, not int seconds per token

Reply:
fixed in latest commit

Reply:
I think it still should be total time not per-token time though right?

Reply:
fixed in latest commit: `the cumulative time of all decode steps in seconds`